src/shared/efi-api.c

```diff
@@ -503,6 +503,12 @@ bool efi_has_tpm2(void) {
         if (errno != ENOENT)
                 log_debug_errno(errno, "Unable to test whether /sys/firmware/acpi/tables/TPM2 exists, assuming it doesn't: %m");
 
+        cache = access("/proc/device-tree/tpm-event-log", F_OK) >= 0;
```

Comment:
This is not standardized and we are better off avoiding it. 
This DT entry comes from TF-A [0] only for QEMU and is not enabled by default.
Since BL2 doesn't have access to TPM drivers, it instead creates an EventLog and expects BL32/33 to replay it.
That entry holds the address of the EventLog


[0] https://github.com/ARM-software/arm-trusted-firmware/blob/master/plat/qemu/qemu/qemu_helpers.c#L41

Reply:
Thanks @apalos. I don't know how to parse devicetree in systemd userspace pid 1 side. There doesn't seem to be kernel interfaces for this until the correct TPM devices have been loaded. Closing this MR.

I noticed at least on arm64 machine, AVA, which reports a TPM device in ACPI but it's not working with our kernel drivers currently.

Reply:
@mikkorapeli-linaro I am not too familiar with systemd internal APIs but the tpm node name is standardized in the DT-bindings [0] and should match '^tpm(@[0-9a-f]+)?$'

The QEMU entry looks like this
/proc/device-tree/platform-bus@c000000/tpm_tis@0/

Could you perhaps 'grep' for a tpm and then look at the compatible or the name? e.g 

```
$~ cat /proc/device-tree/platform-bus@c000000/tpm_tis@0/name@0/name
tpm_tis
$~ cat /proc/device-tree/platform-bus@c000000/tpm_tis@0/compatible 
tcg,tpm-tis-mmio
```

[0] https://www.kernel.org/doc/Documentation/devicetree/bindings/tpm/tpm-common.yaml

Reply:
urgh, iterating through device trees to find a node is a bit much for a quick feature check like we need it. Ideally we'd have a single access() check only like for the ACPI node.

Isn't there any file/dir in sysfs, procfs or elsewhere that either exists on such arm platforms or does not exist if the firmware is providing TPM measured boot? there must be something?

Reply:
I mean, iterating through the tree is not totally impossible, we do have feature check code that looks for virtio devices by scanning through parts of sysfs. But it's a last-resort thing really, it would be *way* better if we had a single flag file we can look for that we don't have to scan for.

Reply:
Problem with this /proc interface is that it's not a single file or directory. Traversing the tree for file matching a regexp is not nice in early boot fast path. I think there should be a better interface for early usespace to detect firmware TPM support via devicetree. I think the ACPI table check for server grade HW is enough for now, though there can be HW where this info is wrong and firmware and/or TPM device is actually broken or kernel side support is missing, like for AVA in our setup currently. For embedded boards, it feels ok to require drivers to be built into the kernel if there is no ACPI table entry. Not sure exactly what the Arm SystemReady variants require from certified HW. https://www.arm.com/architecture/system-architectures/systemready-certification-program

Reply:
> I mean, iterating through the tree is not totally impossible, we do have feature check code that looks for virtio devices by scanning through parts of sysfs. But it's a last-resort thing really, it would be _way_ better if we had a single flag file we can look for that we don't have to scan for.
> 
> (See src/core/kmod-setup.c has_virtio_feature() for an example how to do that)

It depends on when we expect systemd to run and detect all that.
There is /sys/kernel/security/tpm0/binary_bios_measurements. We could either scan for the folder or the file. 
The file though is the EventLog and might not always be there. It also depends on the tpm being built-in or a loaded module.

Reply:
/sys/kernel/security/tpm0/binary_bios_measurements is not available before kernel modules for the TPM device are loaded.